optuna/integration/lightgbm_tuner/optimize.py

```diff
@@ -0,0 +1,608 @@
+from optuna.integration.lightgbm_tuner.alias import _handling_alias_parameters
+from optuna import type_checking
+
+
+if type_checking.TYPE_CHECKING:
+    from type_checking import Any  # NOQA
+    from type_checking import Callable  # NOQA
+    from type_checking import Dict  # NOQA
+    from type_checking import Generator  # NOQA
+    from type_checking import List  # NOQA
+    from type_checking import Optional  # NOQA
+    from type_checking import Tuple  # NOQA
+    from type_checking import Union  # NOQA
+
+    from optuna.distributions import BaseDistribution  # NOQA
+    from optuna.structs import FrozenTrial  # NOQA
+    from optuna.study import Study  # NOQA
+    from optuna.trial import Trial  # NOQA
+
+    VALID_SET_TYPE = Union[List[lgb.Dataset], Tuple[lgb.Dataset, ...], lgb.Dataset]
+
+# Default time budget for tuning `learning_rate`.
+DEFAULT_TIME_BUDGET_FOR_TUNING_LR = 4 * 60 * 60
+
+
+class _GridSamplerUniform1D(optuna.samplers.BaseSampler):
+
+    def __init__(self, param_name, param_values):
+        # type: (str, Any) -> None
+
+        self.param_name = param_name
+        self.param_values = tuple(param_values)
+        self.value_idx = 0
+
+    def sample_relative(self, study, trial, search_space):
+        # type: (Study, FrozenTrial, Dict[str, BaseDistribution]) -> Dict[str, float]
+
+        # todo (g-votte): Take care of distributed optimization.
+        assert self.value_idx < len(self.param_values)
+        param_value = self.param_values[self.value_idx]
+        self.value_idx += 1
+        return {self.param_name: param_value}
+
+    def sample_independent(self, study, trial, param_name, param_distribution):
+        # type: (Study, FrozenTrial, str, BaseDistribution) -> None
+
+        raise ValueError(
+            'Suggest method is called for an invalid parameter: {}.'.format(param_name))
+
+    def infer_relative_search_space(self, study, trial):
+        # type: (Study, FrozenTrial) -> Dict[str, BaseDistribution]
+
+        distribution = optuna.distributions.UniformDistribution(-float('inf'), float('inf'))
+        return {self.param_name: distribution}
+
+
+class _TimeKeeper(object):
+    def __init__(self):
+        # type: () -> None
+
+        self.time = time.time()
+
+    def elapsed_secs(self):
+        # type: () -> float
+
+        return time.time() - self.time
+
+
+@contextlib.contextmanager
+def _timer():
+    # type: () -> Generator[_TimeKeeper, None, None]
+
+    timekeeper = _TimeKeeper()
+    yield timekeeper
+
+
+class BaseTuner(object):
+    def __init__(
+            self,
+            lgbm_params=None,
+            lgbm_kwargs=None
+    ):
+        # type: (Dict[str, Any], Dict[str,Any]) -> None
+
+        self.lgbm_params = lgbm_params or {}
+        self.lgbm_kwargs = lgbm_kwargs or {}
+
+    def _get_booster_best_score(self, booster):
+        # type: (lgb.Booster) -> float
+
+        metric = self.lgbm_params.get('metric', 'binary_logloss')
+
+        # todo (smly): Make this better.
```

Comment:
You may want to leave more specific todo.

Reply:
Fixed: b6e6082